### PR TITLE
Handle empty response 200s

### PR DESCRIPTION
**Pull request by frankiebee**

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -55,3 +57,23 @@ test('fetchConfigFromReq - strip non-standard keys', (t) => {
...
+  const scope = nock('https://api.infura.io').persist().get(/.*/).times(5).reply(200, {})
+  const network = 'mainnet'
+  const req = {
+    method: 'eth_getBlockByNumber',
+    params: ['0x482103', true],
+  }
+
+  const createInfuraMiddleware = require('../src/index')
+
+  const fetchMiddlewear = createInfuraMiddleware()
```
Comment (brunobar79) [inline test/index.js]: typo: fetchMiddleware

File: test/index.js [hunk 2]
```diff
@@ -55,3 +57,23 @@ test('fetchConfigFromReq - strip non-standard keys', (t) => {
...
+
+    fetchMiddlewear(req, {}, () => {}, (e) => {
+      t.true(scope.isDone(), 'should fetch 5 times')
+      t.true(e.message.includes('All retries exhausted'), 'error message should include: All retries exhausted')
+      t.true(e.message.includes('Response has no error or result'), 'error message should include: Response has no error or result')
+      t.end()
+    })
+})
```
Comment (brunobar79) [inline test/index.js]: add new line

File: test/index.js (hunk 1 above)
Comment (frankiebee) [reply test/index.js]: woops sorry thanks!

File: src/index.js [hunk 3]
```diff
@@ -94,6 +96,10 @@ async function performFetch(network, req, res){
   // parse JSON
   const data = JSON.parse(rawData)
 
+  if (!data.result && !data.error) {
```
Comment (danjm) [inline src/index.js]: There are some infura endpoints that can intentionally return a `false` result: https://infura.io/docs/ethereum/json-rpc/eth_mining, https://infura.io/docs/ethereum/json-rpc/eth_submitWork, https://infura.io/docs/ethereum/json-rpc/eth_syncing

Should we do `!(data.result === null || data.result === undefined)` or similar?

**Comment by danjm** [comment]
For a number of Infura's APIs, `null` is a valid response. For example, the result field of https://infura.io/docs/ethereum/json-rpc/eth_getTransactionByHash is "A transaction object, or null when no transaction was found"

Should we be throwing an error in these cases? A user of `eth-json-rpc-infura` who makes a request like `eth_getTransactionByHash` with a hash for which no transaction exists might want to just get the `null` result.

File: src/index.js [hunk 4]
```diff
@@ -94,6 +96,10 @@ async function performFetch(network, req, res){
   // parse JSON
   const data = JSON.parse(rawData)
 
+  if (!('result' in data) && !('error' in data)) {
```
[the remark this answers is not in the document]
Comment (danfinlay) [reply src/index.js]: We could change it to:
`if (!result || !('result' in data) && !('error' in data)) {`

File: test/index.js [hunk 5]
```diff
@@ -55,3 +57,23 @@ test('fetchConfigFromReq - strip non-standard keys', (t) => {
   t.end()
 
 })
+
+test('createInfuraMiddleware - runs through all the retry attempts due to empty response', (t) => {
```
Comment (danfinlay) [inline test/index.js]: We need a test for when `null` is returned by Infura, b/c that's what's breaking this.

File: src/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (frankiebee) [reply src/index.js]: i’d rather not do the result null right now because sometimes result can be null i’m trying to solve the no result exiting the retry loop

**Comment by kumavis** [comment]
we handle the null result elsewhere in some cases
https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/network/createInfuraClient.js#L24
https://github.com/MetaMask/eth-json-rpc-middleware/blob/master/retryOnEmpty.js#L45-L52

**Comment by kumavis** [comment]
the `eth-json-rpc-infura` module is just the wrapper around fetching the infura api

we should publish and write tests the `eth-infura-client` (basically the content of this file https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/network/createInfuraClient.js)
